android/tangram/src/com/mapzen/tangram/MapController.java

```diff
@@ -179,11 +176,22 @@ void init() {
         // Parse font file desription
         fontFileParser.parse();
 
-        nativeInit(this, assetManager, scenePath);
+        mapPointer = nativeInit(this, assetManager);
+    }
+
+    void dispose() {
+        // Disposing native resources involves GL calls, so we need to run on the GL thread.
+        queueEvent(new Runnable() {
+            @Override
+            public void run() {
+                nativeDispose(mapPointer);
+                mapPointer = 0;
```

Comment:
Hmm that's an interesting point. We could clear the `HashMap` of data sources, but the client application could still be holding references and so they could still be used. I think for the sake of producing useful errors in `MapController$AddDataLayer()` it probably does make sense to clear the client data sources here.